File: jsimple-oauth/c#/jsimple/oauth/model/Parameter.cs
using System;

namespace jsimple.oauth.model {

    using OAuthEncoder = jsimple.oauth.utils.OAuthEncoder;


    /// <summary>
    /// @author: Pablo Fernandez
    /// </summary>
    public class Parameter : IComparable<Parameter> {
        private const string UTF = "UTF8";

        private readonly string key;
        private readonly string value;

        public Parameter(string key, string value) {
            this.key = key;
            this.value = value;
        }

        public virtual string asUrlEncodedPair() {
            return OAuthEncoder.encode(key) + "=" + OAuthEncoder.encode(value);
        }

        public override bool Equals(object other) {
            if (other == null)
                return false;
            if (other == this)
                return true;

            if (!(other is Parameter))
                return false;
            Parameter otherParam = (Parameter) other;
            return otherParam.key.Equals(key) && otherParam.value.Equals(value);
        }

        public override int GetHashCode() {
            return key.GetHashCode() + value.GetHashCode();
        }

        public virtual int CompareTo(Parameter parameter) {
            int keyDiff = key.CompareTo(parameter.key);

            return keyDiff != 0 ? keyDiff : value.CompareTo(parameter.value);
        }
    }

}

File: jsimple-oauth/src/main/java/jsimple/oauth/model/Parameter.java
/*
 * Portions copyright (c) 2012-2014 Microsoft Mobile.  All Rights Reserved.
 *
 * This file is based on or incorporates material from the Scribe Java OAuth
 * library https://github.com/fernandezpablo85/scribe-java (collectively, "Third
 * Party Code"). Microsoft Mobile is not the original author of the Third Party
 * Code.
 *
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package jsimple.oauth.model;

import jsimple.oauth.utils.OAuthEncoder;
import org.jetbrains.annotations.Nullable;

/**
 * @author: Pablo Fernandez
 */
public class Parameter implements Comparable<Parameter> {
    private static final String UTF = "UTF8";

    private final String key;
    private final String value;

    public Parameter(String key, String value) {
        this.key = key;
        this.value = value;
    }

    public String asUrlEncodedPair() {
        //return OAuthEncoder.encode(key).concat("=").concat(OAuthEncoder.encode(value));
        return OAuthEncoder.encode(key) + "=" + OAuthEncoder.encode(value);
    }

    @Override public boolean equals(@Nullable Object other) {
        if (other == null)
            return false;
        if (other == this)
            return true;

        if (!(other instanceof Parameter))
            return false;
        Parameter otherParam = (Parameter) other;
        return otherParam.key.equals(key) && otherParam.value.equals(value);
    }

    @Override public int hashCode() {
        return key.hashCode() + value.hashCode();
    }

    public int compareTo(Parameter parameter) {
        int keyDiff = key.compareTo(parameter.key);

        return keyDiff != 0 ? keyDiff : value.compareTo(parameter.value);
    }
}
